### PR TITLE
Blog - deploying a monolith rails application

### PR DESCRIPTION
**Pull request by NgariNdungu**
[related file](https://github.com/NgariNdungu/zegetech.github.io/blob/d5dfbb0c4fa48c3b6c9240d6d4d76dae7f8c29d0/_posts/2018-11-13-deploy-rails.md)
[updated](https://github.com/NgariNdungu/zegetech.github.io/blob/feature-blog-deploy-monolith/_posts/2018-11-13-deploy-rails.md)

### CODE REVIEW

File: _posts/2018-11-13-deploy-rails.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+blog-image: 
+intro: So, you finally have your app in a state that you aren't too embarrassed to show off. How do you open it up to the world?
+  How do you manage the influx of users? How will you keep it safe from the *dangerous* internets? These are questions we face every time we work on a new product.
+  This post explores one of the ways to deploy your shiny new rails application and the common tooling behind it.
+---
+
+{{ page.intro }}
+
+## Monolithic?
+
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md:18]: Mention somewhere that it is good practice not to wait till you application is done to deploy it, but to start with a deployed boilerplate, in rails case, deploying a rails new app, and making user the changes are also always deployed.. Also mention the need for staging, production and dev environments.

File: _posts/2018-11-13-deploy-rails.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+That's the typical MVC framework setup.
+
+You will have heard of micro-services in the last few years. The micro-service architecture encourages the development of applications as disparate services, with each service catering to a single business requirement.
+Each service can be viewed as a monolith application, albeit with reduced business scope.
+We'll cover deploying a *traditional* rails app in this post, and follow it up with deploying a micro-services based app in a later post.
+
+## Deployment options
+
+Unless you are lucky enough to work in a server room and have one of those servers as your development machine, you will need somewhere to host your site.
+We have the choice of deploying our site to a PaaS provider such as [Heroku] or using a VPS from a IaaS provider.
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: Mention cloud and link to the article about cloud

File: _posts/2018-11-13-deploy-rails.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+Deploying to a platform such as heroku abstracts away the complexities of setting up a performant and secure environment for your application.
+No need to worry about setting up and updating the OS, the database server or any of your app dependencies. You simply upload your code and the platform takes care of the rest.
+
+The more *painful* approach is to create a VPS on a provider such as [digital ocean], [linode] or [aws lightsail].
+With a vps, you control every aspect of the server. If something fails, that's on you.
+The process of setting up is not much different from what you did to get the app running on your development machine.
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md:61]: Please mention the advantages over easier PaaS, like control over dependancies, services as well as huge cost savings and optimisation

File: _posts/2018-11-13-deploy-rails.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+But, on top of getting your app to run, you have to make considerations for overall site security and app performance.
+And you guessed it... we're going with the painful approach.
+
+## Our production environment
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: Again, mention as best practice, you need 2 to 3  environments for a real live system. development, staging(bleeding edge) and production(stable)

File: _posts/2018-11-13-deploy-rails.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+### Base Environment
+
+You will mostly get your pick of operating system with most cloud providers. And here is where the choices start.
+We will be going with Ubuntu 18.04(Bionic Beaver). This is an LTS(long term support) version of Ubuntu with at least 5 years of security support.
+Unless your deploy will be short lived, make sure to choose an OS with guaranteed updates.
+
+#### User Setup
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: Divide your sections into steps as this is a practical blog that you want devs to follow through

File: _posts/2018-11-13-deploy-rails.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+~~~shell
+rails new [appname] -m /path/to/template.rb # path can also be a url
+~~~
+Or to apply to an existing app:
+~~~shell
+bin/rails app:template LOCATION=/path/to/template.rb # location variable can also be a url
+~~~
+
+#### Example you ask?
+
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md:183]: add a relevant image to draw attention. also include the first steps that help people to get going with rails_seed.

Comment (zegetech) [reply _posts/2018-11-13-deploy-rails.md:183]: Mention that rails seed is an app template that we developed at zegetech

File: _posts/2018-11-13-deploy-rails.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+[Rails seed](https://github.com/kgathi2/rails_seed) is a Rails Application Template with the lofty goal of getting you up and running a production-ready environment in 15 minutes.
+And it works... mostly. If nothing else, it should give you a more [concrete](https://sw.wikipedia.org/wiki/Zege) image of the gibberish written here.
+The template contains tasks that take care of installing the typical ruby and rails dependencies, setting up postgres, adding firewall rules and most other setup tasks.
+
+The template for now works for Rails 5.0 and is tested on Ubuntu xenial and bionic. It should also work with rails >=5.1 once you [setup encrypted credentials](https://www.starkandwayne.com/blog/rails-5-1-applications-can-be-a-lot-more-secretive-on-cloud-foundry-and-heroku/).
+
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md:248]: Include "how to" of deploying rails seed on local without using a VPS in the cloud. Talking about vagrant and docker. So basically a step by step for how to use the template to deploy to a docker container or vagrant.

**Review by zegetech — CHANGES_REQUESTED** [review]

**Comment by NgariNdungu** [comment]
I have made the requested changes. [Here's the updated post](https://github.com/NgariNdungu/zegetech.github.io/blob/feature-blog-deploy-monolith/_posts/2018-11-13-deploy-rails.md)

File: _posts/2018-11-13-deploy-rails.md (hunk 5 above)
Comment (zegetech) [reply _posts/2018-11-13-deploy-rails.md]: Please use number that are easier to follow in addition to H1 H2 etc...

File: _posts/2018-11-13-deploy-rails.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+Unless your deploy will be short lived, make sure to choose an OS with guaranteed updates.
+
+#### User Setup
+
+When you create a new server with the cloud providers defaults, you mostly receive a root password in your email.
+You can then log in to the server as the root user. This base setup is insecure. If your password was to fall into the wrong hands, your server would no longer be yours!
+Running as the root user also does not leave room for error. There will be no prompts warning you when you try to remove a critical system file.
+
+It is recommended practice to always create a *normal* user and use SSH key authentication to log in to your server.
+You can refer to [this article](https://www.digitalocean.com/community/tutorials/initial-server-setup-with-ubuntu-16-04) for an explanation and step-by-step.
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: deploy user or  user for deploying

File: _posts/2018-11-13-deploy-rails.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+An alternative to performing these steps after your server has been built is to provide a script that the cloud provider will run as part of the provisioning task.
+This is referred as *User data* on digital ocean while aws lightsail asks for a *launch script*. Here's an [example script](https://github.com/kgathi2/rails_seed#vps-preparation).
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: Mention cloud_init which user data works of off.

File: _posts/2018-11-13-deploy-rails.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+You can also provide your SSH public key to completely skip setting up password login to the server.
+
```
Comment (zegetech) [inline _posts/2018-11-13-deploy-rails.md]: Add how to set up ssh key on local host for ssh authentication without password

**Review by zegetech — CHANGES_REQUESTED** [review]

**Review by zegetech — APPROVED** [review]

**Pull request merged**